exchanges/okx/okx_wrapper.go

```diff
@@ -0,0 +1,1134 @@
+				CancelOrders:        true,
+				TradeFetching:       true,
+				UserTradeHistory:    true,
+			},
+			WebsocketCapabilities: protocol.Features{
+				TickerFetching:         true,
+				OrderbookFetching:      true,
+				Subscribe:              true,
+				Unsubscribe:            true,
+				AuthenticatedEndpoints: true,
+				AccountInfo:            true,
+				GetOrders:              true,
+				TradeFetching:          true,
+				KlineFetching:          true,
+				GetOrder:               true,
+			},
+			WithdrawPermissions: exchange.AutoWithdrawCrypto,
+		},
+		Enabled: exchange.FeaturesEnabled{
+			AutoPairUpdates: true,
+			Kline: kline.ExchangeCapabilitiesEnabled{
+				Intervals: map[string]bool{
+					kline.OneMin.Word():     true,
+					kline.ThreeMin.Word():   true,
+					kline.FiveMin.Word():    true,
+					kline.FifteenMin.Word(): true,
+					kline.ThirtyMin.Word():  true,
+					kline.OneHour.Word():    true,
+					kline.TwoHour.Word():    true,
+					kline.FourHour.Word():   true,
+					kline.SixHour.Word():    true,
+					kline.TwelveHour.Word(): true,
+					kline.OneDay.Word():     true,
+					kline.ThreeDay.Word():   true,
+					kline.OneWeek.Word():    true,
+					kline.OneMonth.Word():   true,
+					kline.ThreeMonth.Word(): true,
+					kline.SixMonth.Word():   true,
+					kline.OneYear.Word():    true,
+				},
+				ResultLimit: 1440,
+			},
+		},
+	}
+	ok.Requester, err = request.New(ok.Name,
+		common.NewHTTPClientWithTimeout(exchange.DefaultHTTPTimeout),
+		request.WithLimiter(SetRateLimit()))
+	if err != nil {
+		log.Errorln(log.ExchangeSys, err)
+	}
+
+	ok.API.Endpoints = ok.NewEndpoints()
+	err = ok.API.Endpoints.SetDefaultEndpoints(map[exchange.URL]string{
+		exchange.RestSpot:      okxAPIURL,
+		exchange.WebsocketSpot: okxAPIWebsocketPublicURL,
+	})
+	if err != nil {
+		log.Errorln(log.ExchangeSys, err)
+	}
+
+	ok.Websocket = stream.New()
+	ok.WebsocketResponseMaxLimit = okxWebsocketResponseMaxLimit
+	ok.WebsocketResponseCheckTimeout = okxWebsocketResponseMaxLimit
+	ok.WebsocketOrderbookBufferLimit = exchange.DefaultWebsocketOrderbookBufferLimit
+}
+
+// Setup takes in the supplied exchange configuration details and sets params
+func (ok *Okx) Setup(exch *config.Exchange) error {
+	err := exch.Validate()
+	if err != nil {
+		return err
+	}
+	if !exch.Enabled {
+		ok.SetEnabled(false)
+		return nil
+	}
+	err = ok.SetupDefaults(exch)
+	if err != nil {
+		return err
+	}
+
+	wsRunningEndpoint, err := ok.API.Endpoints.GetURL(exchange.WebsocketSpot)
+	if err != nil {
+		return err
+	}
+	err = ok.Websocket.Setup(&stream.WebsocketSetup{
+		ExchangeConfig:        exch,
+		DefaultURL:            okxAPIWebsocketPublicURL,
+		RunningURL:            wsRunningEndpoint,
+		Connector:             ok.WsConnect,
+		Subscriber:            ok.Subscribe,
+		Unsubscriber:          ok.Unsubscribe,
+		GenerateSubscriptions: ok.GenerateDefaultSubscriptions,
+		Features:              &ok.Features.Supports.WebsocketCapabilities,
```

Comment:
You can actually pass in the checksum function:

```
OrderbookBufferConfig: buffer.Config{
			Checksum: ok.CalculateUpdateOrderbookChecksum,
		},
```

This is slightly more efficient at validating books because it saves on mutex lock calls.